### ISSUE TITLE
HTTP proxy implementation does not escape correctly

### ISSUE THREAD

**Issue by dagwieers** [openshift/origin#11338]
$CUSTOMER has to use an HTTP proxy to access the internet. Authentication requires an Active Directory account in the format: DOMAIN\USER or USER@DOMAIN which means the URL to the proxy results in:

```
http://DOMAIN\USER:password@proxy.local:8080/
```

or

```
http://USER@DOMAIN:password@proxy.local:8080/
```

Both are problematic for various reasons:
- Docker and Openshift require environment variables in a shell-script, which means the backslash needs to be escaped ! (Which it does not, so we have to do it ourselves)
  - openshift_http_proxy='http://DOMAIN\USER:password@proxy.local:8080/'
  - openshift_https_proxy='http://DOMAIN\USER:password@proxy.local:8080/'
- Openshift provides the proxy settings for pulling sources and pushing images, and this leads to various problems depending on the tools that require these settings, e.g.
  - github works fine with the backslash as is (so it should not be escaped)
  - however for npm there is an issue because the backslash is being translated into %5C, and npm is trying to resolve DOMAIN rather than proxy.local.

So the whole http_proxy implementation is very prone to issues specific to how this specific string is being used by individual tools.
##### Version

Openshift Container Platform 3.3
##### Steps To Reproduce
1. Configure the HTTP_PROXY in your Ansible inventory and use an Active Directory style username
2. Install/re-configure your setup
3. Watch it fail to build or push (or worse, fail to see any proper logs)
   ##### Current Result

Failure
##### Expected Result

Success

**Comment by dagwieers** [comment]
Example output from builds:

```
Using HTTP proxy http://USER%40DOMAIN:password@proxy.local:8080/ and HTTPS proxy http://USER%40DOMAIN:password@proxy.local:8080/ for script download
Cloning "https://github.com/openshift/cakephp-ex.git" ...
error: build error: fatal: unable to access 'https://github.com/openshift/cakephp-ex.git/'; Could not resolve proxy: DOMAIN; Name or service not known
```

```
---> Installing application source ...
---> Setting nm http proxy to http://DOMAIN\USER:password@proxy.local:8080/
---> Setting nm http proxy to http://DOMAIN\USER:password@proxy.local:8080/
---> Setting nm https proxy to http://DOMAIN\USER:password@proxy.local:8080/
---> Setting nm https proxy to http://DOMAIN\USER:password@proxy.local:8080/
---> Building your Node application from Source
npm info it worked if it ends with ok
npm info using npm@2.15.1
npm info using node@v4.4.2
npm info preinstall nodejs-ex@0.0.1
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/ejs
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/express
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/mocha
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/mongodb
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/morgan
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/object-assign
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/chai
npm info attempt registry request try #1 at 10:26:56
npm http request GET https://registry.npmjs.org/chai-http
npm info retry will retry, error on last attempt: Error: tunneling socket could not be establishd, cause=getaddrinfo ENOTFOUND domain domain:80

**Comment by dagwieers** [comment]
BTW It would be very useful if one could set/override the NO_PROXY, HTTP_PROXY and HTTPS_PROXY environment variables oneself via the Builds-interface in the Environment tab, so that one can set these variables on a per-build/per-project, even when it is configured globally in Openshift.

**Comment by dagwieers** [comment]
Related links:
- http://stackoverflow.com/questions/25660936/using-npm-behind-corporate-proxy-pac

**Comment by dagwieers** [comment]
And on top of this issue the push-stage does not take into account the NO_PROXY or no_proxy variable. Whether using a network-range (CIDR) or a specific IP address, it fails to match and nevertheless uses the proxy (when it shouldn't). Very frustrating since there is no combination of settings that make everything work as it should :-(

**Comment by sdodson** [comment]
> And on top of this issue the push-stage does not take into account the NO_PROXY or no_proxy variable.

The push stage uses docker on the host, so make sure /etc/sysconfig/docker has NO_PROXY value with the service IP of the docker registry. We're trying to move towards pushing via dns name (ie: docker-registry.default.svc.cluster.local) but there are various pieces that had to be fixed across the product to enable that. Docker doesn't support NO_PROXY CIDR ranges, see https://github.com/docker/docker/issues/9145

**Comment by sdodson** [comment]
@bparees Any thoughts on which parts of this should be solved via s2i tooling? If this varies based on the builder image I think that's the only place we could really handle it. If it's universal across all of them then I guess I could teach the installer to escape things differently for /etc/sysconfig/{origin-node,origin-master,docker} and the values set in the buildDefaults admission controller.

**Comment by dagwieers** [comment]
@sdodson So our internal registry is 172.30.224.67 and the NO_PROXY is automatically stuff with 172.30.0.0/16 which should make it work, but it doesn't. We tried adding 172.30.0.0/12 and also 172.30.224.67 itself. No difference, it appears as if pushing doesn't really take NO_PROXY into account.

This is on Red Hat Openshift Container Platform v3.3. 

Very frustrating to try to make the best out of a proof-of-concept but getting stuck on proxy issues all over the place. We cannot even use any of the examples because either they don't build, they fail to get pushed, or worse most of the example images are not being pulled due to weird proxy issues. But the installer doesn't complain, and command report operations went successfully, but then also report various weird authentication errors (which I assume are related to our proxy ?)

**Comment by dagwieers** [comment]
Some tickets have been opened with Red Hat Support BTW.

**Comment by dagwieers** [comment]
> The push stage uses docker on the host, so make sure /etc/sysconfig/docker has NO_PROXY value with the service IP of the docker registry.
Check!

> We're trying to move towards pushing via dns name (ie: docker-registry.default.svc.cluster.local) but there are various pieces that had to be fixed across the product to enable that. Docker doesn't support NO_PROXY CIDR ranges, see docker/docker#9145
Well, the problem we have with NO_PROXY is not related to docker, but related to builder/push.go and I looked at the code and it does seem to support it. It might still be a bug, but support is specifically implemented.

**Comment by dagwieers** [comment]
And it also seems as if updating proxy settings does not immediately affect all services because I assume some are not being restarted. Ironically that was the only situation where we at some point had a working situation (except for pushing the image), first use a working solution for one part of Openshift, then reboot, then use another config that does get activated. And then it works until you reboot the node(s).

**Comment by bparees** [comment]
I'm not sure how any of this can easily be solved by s2i tooling, you're really talking about needing specific proxy values/formats depending on what s2i builder image you're using.  The fact is global defaults can only get you so far.

**Comment by dagwieers** [comment]
@bparees Not at all. We have a global default that should theoretically work for everything. Unfortunately, every single tool that should be using our variables handles them differently.

- npm parse the URL at the first @ it finds (however a username could include a @, so with USER@DOMAIN it thinks the DOMAIN is the host-part, see example)
- npm also does not support a backslash in the username (AD uses DOMAIN\USER as username)
- builder/push.go does not support NO_PROXY apparently, because whatever we throw at it, it always goes to the proxy, not to our integrated repository
- The proxy variables in a shell script (/etc/sysconfig) should escape the backslash
- Tools like git should not need an escaped backslash
- docker apparently does not work with CIDR inside NO_PROXY
- and some other weird things we noticed during experiments...

It doesn't help that some tools do not provide proper information to why it fails, and it also does not help that I do not have access to the proxy to check why it fails (or sometimes times out).

**Comment by dagwieers** [comment]
And BTW the default NO_PROXY content that the installer adds should be working out-of-the-box. We only added stuff to try and work around some of the above. With little success :-/ We could get some things working by modifying files manually and restarting services, but we cannot make both pull and push work together because push uses the proxy when it shouldn't.

**Comment by sdodson** [comment]
Do you see signs of failed pushes in the journal for the docker process? If you see them there then the builder is properly triggering the push but the daemon cannot push to the service ip, which is what i'd expect. Though if that were the case I'd expect pushes and pulls to fail. When you say pulls work is that a pull from the integrated registry which would fail if NO_PROXY weren't working properly or is that from an external host which should actually use the proxy?

**Comment by dagwieers** [comment]
I am now seeing this BTW
```
Pushing image 172.30.224.67:5000/dag/nodejs-example:latest ...
error: build error: Failed to push image: no basic auth credentials
```

Where before we used to get:
```
Pushing images 172.30.224.67:5000/dag/nodejs-example:latest ...
error: build error: Failed to push image: Error: Status 407 trying to push repository dag/nodejs-example: "<!DOCTYPE html PUBLIC \"-//W3C/DTD XHTML 1.0 Transitional//EN\" \http://www.w3.org/TR/xhtml11/DTD/xhtml11-transitional.dtd\">\n<html>\n<!-- FileName: index.html\n     Language: [en]\n-->\n<!--Head-->\n<head>\n...
```
Which definitely was output from the proxy. Now that this is gone, maybe the NO_PROXY with a specific IP address does seem to work ? Any ideas ?

**Comment by dagwieers** [comment]
BTW Are there any neat tricks to debug what is going on in a container when it fails ? Like increasing verbosity, keep the container alive, debugging the network from the exterior, etc.

**Comment by dagwieers** [comment]
The pull's that work seem to be Git pulls. But for some new builds from the examples it fails because the Build Image it expects were never provisioned (I assume) due to proxy errors. I opened a ticket related to that to Red Hat Support.

**Comment by sdodson** [comment]
https://docs.openshift.com/enterprise/3.2/dev_guide/builds.html#accessing-build-logs describes how to increase the log level for builds

**Comment by dagwieers** [comment]
Just FYI I got another error when doing the dancer-example, after fetching the first module "App:cpanminus" it bails out with **gzip: stdin: unexpected en of file**. Most likely another victim of the proxy configuration :-/

**Comment by dagwieers** [comment]
And to show you I am not getting nuts, output before and after a reboot:

**Before reboot**
```
Using HTTP proxy http://USER:PASSWORD@SERVER:8080/ and HTTPS proxy http://USER:PASSWORD@SERVER:8080/ for script download
Cloning "https://github.com/openshift/nodejs-ex.git" ...
        Commit: 9457131bdbd399d1242ec7f4b2fa4fd4fcb5b05b (Adds NodeJS/MongoDB template with persistent storage. (#64))
        Author: Jacob Lucky <jlucky@redhat.com>
        Date: Tue Sep 20 09:12:04 2016 -0400

---> Installing application source ...
---> Setting npm http proxy to http://USER:PASSWORD@SERVER:8080/
---> Setting npm http proxy to http://USER:PASSWORD@SERVER:8080/
---> Setting npm https proxy to http://USER:PASSWORD@SERVER:8080/
---> Setting npm https proxy to http://USER:PASSWORD@SERVER:8080/
---> Building your Node application from source
npm info it worked if it ends with ok
npm info using npm@2.15.1
npm info using node@v4.4.2
npm info preinstall nodejs-ex@0.0.1
...
```

**After reboot**
```
Using HTTP proxy http://http://USER:PASSWORD@SERVER:8080/ and HTTPS proxy http://USER:PASSWORD@SERVER:8080/ for script download
Cloning "https://github.com/openshift/nodejs-ex.git" ...
error: build error: fatal: unable to access 'https://github.com/openshift/nodejs-ex.git/': Received HTTP code 407 from proxy after CONNECT
```

So while the output seems to be using the exact same proxy configuration, the behaviour is clearly different. And I have no clue why that would be.

**Comment by dagwieers** [comment]
Alright, $COMPANY disabled proxy-authentication for our IP addresses, so we don't need to provide a username and password. And everything works as expected.

So I can conclude that OpenShift has an issue with proxy authentication (as we suspected) and handling NO_PROXY. (CIDR are not supported, but are nevertheless added during installation) Issues are in the various specific tools used for downloading sources.

We still had a problem with pushing images to the docker registry, but that was resolved by adding the docker-registry explicitly to the NO_PROXY list as advised by Red Hat Support. (See openshift/openshift-ansible#1919) *and* by recreating the various projects (I assume a reinstallation changed credentials or keys which caused auth issues)

We still have issues with missing imagestreams (e.g. *imagestreamtags "python:3.5" not found*) and GUI issues adding templates to a project (e.g. *Cannot create from source: the specified image could not be retrieved.*) that do not seem directly related to proxy-settings.
